cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -379,6 +457,109 @@ writeLedgerState mOutFile qState@(SerialisedLedgerState serLedgerState) =
+                   -> SerialisedLedgerState era
+                   -> ExceptT ShelleyQueryCmdError IO ()
+
+writeStakeSnapshot poolId qState =
+  case decodeLedgerState qState of
+    Left bs -> firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs
+    Right ledgerState ->
+      if isNothing maybehk then
+        left $ ShelleyQueryCmdPoolIdError poolId
+      else
+        liftIO . LBS.putStrLn $  encodePretty $ Stakes {markpool = markStake, setpool = setStake, gopool = goStake, marktot = markTotal, settot = setTotal, gotot = goTotal}
+      where
+        -- Ledger State
+        (LedgerState snapshot) = ledgerState
+
+        -- The three stake snapshots, obtained from the ledger state
+        (SnapShots markS setS goS _) = esSnapshots $ nesEs snapshot
+
+        -- Calculate the three pool and active stake values for the given pool
+        markStake =  getPoolStake hk markS
+        setStake = getPoolStake hk setS
+        goStake = getPoolStake hk goS
+
+        markTotal = getAllStake markS
+        setTotal = getAllStake setS
+        goTotal = getAllStake goS
+
+        -- Convert the hash string into a KeyHash for use by the ledger
+        maybehk = hashFromStringAsHex $ filter (/= '"') $ show poolId
+        hk = KeyHash $ fromJust maybehk
+ where
+   decodeLedgerState
+     :: SerialisedLedgerState era
+     -> Either LBS.ByteString (LedgerState era)
+   decodeLedgerState (SerialisedLedgerState (Serialised ls)) =
+     first (const ls) (decodeFull ls)
+
+   -- Sum all the stake that is held by the pool
+   getPoolStake :: KeyHash Shelley.Spec.Ledger.Keys.StakePool crypto
+                -> SnapShot crypto
+                -> Integer
+   getPoolStake hk ss = pStake
+      where
+        Coin pStake = fold s
+             where
+                (Stake s) = poolStake hk (_delegations ss) (_stake ss)
+
+   -- Sum the active stake from a snapshot
+   getAllStake :: SnapShot crypto
+               -> Integer
+   getAllStake (SnapShot stake _ _) = activeStake
+      where
+        Coin activeStake = fold . unStake $ stake
+
+-- | This function obtains the pool parameters, equivalent to the following jq query on the output of query ledger-state
+--   .nesEs.esLState._delegationState._pstate._pParams.<pool_id>
+writePoolParams :: forall era ledgerera.
+                   ShelleyLedgerEra era ~ ledgerera
+                => FromCBOR (LedgerState era)
+                => Crypto.Crypto (Era.Crypto ledgerera)
+                => Hash StakePoolKey
+                -> SerialisedLedgerState era
+                -> ExceptT ShelleyQueryCmdError IO ()
+writePoolParams poolId qState =
+  case decodeLedgerState qState of
+    Left bs -> firstExceptT ShelleyQueryCmdHelpersError $ pPrintCBOR bs
+    Right ledgerState ->
+      if isNothing maybehk then
+        left $ ShelleyQueryCmdPoolIdError poolId
+      else
+        liftIO . LBS.putStrLn $  encodePretty $ Params poolparams fpoolparams retiring
+      where
```

Comment:
Let's move the functions/values from this where clause to the outer where clause. No need to have two.

Reply:
Apart from `hk`, the definitions depend on the match to  `Right ledgerstate`, so cannot be lifted out.  I have pushed the function definitions into the inner scope since they are not currently used in the Left branch